### PR TITLE
node/approval-voting: Continue to migrate tests to subsystem tests

### PR DESCRIPTION
**Pull request by Lldenaurois**
This PR addresses the current gap between tests.rs and old_tests.rs

The goal is to provide the same amount of coverage using full subsystem tests.

### CODE REVIEW

File: node/core/approval-voting/src/tests.rs [hunk 1]
```diff
@@ -394,7 +448,7 @@ fn test_harness<T: Future<Output = VirtualOverseer>>(
 		ApprovalVotingSubsystem::with_config(
 			Config{
 				col_data: test_constants::TEST_CONFIG.col_data,
-				 slot_duration_millis: 100u64,
+				slot_duration_millis: 5000u64,
```
Comment (rphmeier) [inline node/core/approval-voting/src/tests.rs:451]: Why did this need to change? Time should be totally programmatic in the tests. Also recommend extracting to a constant `TEST_SLOT_DURATION` or something.

File: node/core/approval-voting/src/tests.rs [hunk 2]
```diff
@@ -332,6 +333,60 @@ impl Backend for TestStore {
 	}
 }
 
+#[derive(Clone)]
+pub(super) struct SharedTestStore<T: Backend> {
```
Comment (rphmeier) [inline node/core/approval-voting/src/tests.rs:337]: I wonder if this needs to be generic. After all, in test code every type should be known. Or is this used with multiple different kinds of backends?

Comment (Lldenaurois) [reply node/core/approval-voting/src/tests.rs:321]: If you feel this should still be a specific type, I can address that as well.

File: node/core/approval-voting/Cargo.toml [hunk 3]
```diff
@@ -32,6 +32,7 @@ sp-application-crypto = { git = "https://github.com/paritytech/substrate", branc
 sp-runtime = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
 
 [dev-dependencies]
+async-std = "1.6.5"
```
Comment (ordian) [inline node/core/approval-voting/Cargo.toml:35]: seems it's only used for `async_std::task::sleep` which is an overkill, we can use `futures_timer::Delay` instead

Comment (Lldenaurois) [reply node/core/approval-voting/Cargo.toml:35]: Ah yes, I forgot to remove async-std. Great catch, thank you!

File: node/core/approval-voting/src/lib.rs [hunk 4]
```diff
@@ -86,9 +86,6 @@ use crate::backend::{Backend, OverlayedBackend};
 #[cfg(test)]
 mod tests;
 
-#[cfg(test)]
-mod old_tests;
```
Comment (drahnr) [inline node/core/approval-voting/src/lib.rs:90]: :heart:

File: node/core/approval-voting/src/tests.rs [hunk 5]
```diff
@@ -513,9 +658,64 @@ fn blank_subsystem_act_on_bad_block() {
 }
 
 #[test]
-fn ss_rejects_approval_if_no_block_entry() {
+fn subsystem_rejects_approval_if_no_candidate_entry() {
+	let store = SharedTestStore::new(TestStore::default());
+	let (oracle, _handle) = make_sync_oracle(false);
+	let (mock_clock, assignment_criteria) = make_harness_input();
```
Comment (drahnr) [inline node/core/approval-voting/src/tests.rs:664]: nit: this is rather repetitive, a builder pattern with replacable defaults comes to mind that would reduce this a bit, but hey, it's a test, so feel free to ignore.

File: node/core/approval-voting/src/tests.rs [hunk 6]
```diff
@@ -616,11 +827,15 @@ fn ss_rejects_assignment_in_future() {
 
 		// Add block hash 00.
 		ChainBuilder::new()
-			.add_block(block_hash, ChainBuilder::GENESIS_HASH, Slot::from(1), 1)
+			.add_block(block_hash, ChainBuilder::GENESIS_HASH, 1, BlockConfig {
```
Comment (drahnr) [inline node/core/approval-voting/src/tests.rs:830]: :+1:

File: node/core/approval-voting/src/tests.rs [hunk 7]
```diff
@@ -1280,3 +1703,858 @@ fn forkful_import_at_same_height_act_on_leaf() {
...
+	slot: Slot,
+	tranche_offset: Tick,
+	block_hash: Hash,
+	candidate_hash: CandidateHash,
+	candidate_index: CandidateIndex,
+) -> bool {
+		assert!(clock.inner.lock().has_wakeup(slot_to_tick(slot) + tranche_offset));
+		clock.inner.lock().wakeup_all(slot_to_tick(slot) + tranche_offset);
+
+		async_std::task::sleep(std::time::Duration::from_millis(200)).await;
```
Comment (drahnr) [inline node/core/approval-voting/src/tests.rs:2225]: I'd prefer to craft explicit features,i.e. a oneshot with a timeout to assure synchronization _if_ possible.

File: node/core/approval-voting/src/tests.rs [hunk 8]
```diff
@@ -1280,3 +1703,858 @@ fn forkful_import_at_same_height_act_on_leaf() {
...
+fn subsystem_assignment_not_triggered_more_than_maximum() {
+	triggers_assignment_test(TriggersAssignmentConfig {
+		our_assigned_tranche: 3,
+		assign_validator_tranche: |_| Ok(0),
+		no_show_slots: 2,
+		assignments_to_import: vec![2, 3],
+		approvals_to_import: vec![],
+		ticks: vec![
+			13, // Alice wakeup
+			20, // Check no shows
```
Comment (drahnr) [inline node/core/approval-voting/src/tests.rs:2488]: :+1:

File: node/core/approval-voting/src/tests.rs [hunk 9]
```diff
@@ -1280,3 +1703,858 @@ fn forkful_import_at_same_height_act_on_leaf() {
...
+				vec![ValidatorIndex(3), ValidatorIndex(4)],
+			],
+			needed_approvals: 2,
+			discovery_keys: validators.iter().map(|v| v.public().into()).collect(),
+			assignment_keys: validators.iter().map(|v| v.public().into()).collect(),
+			n_cores: validators.len() as _,
+			zeroth_delay_tranche_width: 5,
+			relay_vrf_modulo_samples: 3,
+			n_delay_tranches: 50,
+			no_show_slots: 2,
```
Comment (drahnr) [inline node/core/approval-voting/src/tests.rs:2180]: nit: A few comments on the intent why they were chosen as they would be nice

**Review by drahnr — COMMENTED** [review]
A few nits, generally looks good to me on initial review pass.

**Review by ordian — APPROVED** [review]
Good to go once you `cargo +nightly fmt` :+1:

**Pull request merged**